ipam/allocator.go

```diff
@@ -159,97 +187,293 @@ const (
+		}
+		goto finalize
+	}
+
+	a.Unlock()
+
+	// parent does not exist
+	pp = &PoolData{
+		Pool:     p.Pool,
+		RefCount: 1,
+	}
+	if err := a.insertBitMask(p.ParentKey, pp.Pool); err != nil {
+		return "", nil, nil, err
+	}
+
+	a.Lock()
+	a.subnets[p.ParentKey] = pp
+finalize:
 	a.Unlock()
 	err = a.writeToStore()
 	if err != nil {
 		if _, ok := err.(types.RetryError); !ok {
-			return types.InternalErrorf("subnet configuration failed because of %s", err.Error())
+			return "", nil, nil, types.InternalErrorf("pool configuration failed because of %s", err.Error())
 		}
 		// Update to latest
 		if erru := a.readFromStore(); erru != nil {
-			// Restore and bail out
-			a.Lock()
-			delete(a.addresses, key)
-			a.Unlock()
-			return fmt.Errorf("failed to get updated subnets config from datastore (%v) after (%v)", erru, err)
+			return "", nil, nil, fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
 		}
 		goto retry
 	}
 
-	// Insert respective bitmasks for this subnet
-	a.insertAddressMasks(key, subnetList)
+	return k.String(), p.Pool, nil, nil
+}
+
+func (a *Allocator) insertBitMask(key subnetKey, pool *net.IPNet) error {
+	ipVer := getAddressVersion(pool.IP)
+	ones, bits := pool.Mask.Size()
+	numAddresses := uint32(1 << uint(bits-ones))
+
+	if ipVer == v4 {
+		// Do not let broadcast address be reserved
+		numAddresses--
+	}
+
+	// Generate the new address masks. AddressMask content may come from datastore
+	h, err := bitseq.NewHandle(dsDataKey, a.store, key.String(), numAddresses)
+	if err != nil {
+		return err
+	}
+
+	if ipVer == v4 {
+		// Do not let network identifier address be reserved
+		h.Set(0)
+	}
+
+	a.Lock()
+	a.addresses[key] = h
+	a.Unlock()
 
 	return nil
 }
 
-// Create and insert the internal subnet(s) addresses masks into the address database. Mask data may come from the bitseq datastore.
-func (a *Allocator) insertAddressMasks(parentKey subnetKey, internalSubnetList []*net.IPNet) error {
-	ipVer := getAddressVersion(internalSubnetList[0].IP)
-	num := len(internalSubnetList)
-	ones, bits := internalSubnetList[0].Mask.Size()
-	numAddresses := 1 << uint(bits-ones)
+// ReleasePool releases the address pool identified by the passed id
+func (a *Allocator) ReleasePool(poolID string) error {
+	var (
+		p, pp *PoolData
+		rk    *subnetKey
+		ok    bool
+	)
+	k := subnetKey{}
+	if err := k.FromString(poolID); err != nil {
+		return types.BadRequestErrorf("invalid pool id: %s", poolID)
+	}
 
-	for i := 0; i < num; i++ {
-		smallKey := subnetKey{parentKey.addressSpace, parentKey.subnet, internalSubnetList[i].String()}
-		limit := uint32(numAddresses)
+retry:
+	a.Lock()
+	p, ok = a.subnets[k]
+	if !ok {
+		a.Unlock()
+		return ipamapi.ErrBadPool
+	}
+	p.RefCount--
+	if p.RefCount == 0 {
+		delete(a.subnets, k)
+		rk = &k
+	}
 
-		if ipVer == v4 && i == num-1 {
-			// Do not let broadcast address be reserved
-			limit--
+	if p.Range != nil {
+		pp, ok = a.subnets[p.ParentKey]
```

Comment:
Again, we need to understand this parent key concept. I am not sure we need this complexity